library/std/src/sys/uefi/net/tcp4.rs

```diff
@@ -0,0 +1,473 @@
+    pub(crate) fn reset(&self) -> io::Result<()> {
+        unsafe { Self::config_raw(self.protocol.as_ptr(), crate::ptr::null_mut()) }
+    }
+
+    pub(crate) fn accept(&self) -> io::Result<Tcp4Protocol> {
+        let accept_event = common::Event::create(
+            r_efi::efi::EVT_NOTIFY_WAIT,
+            r_efi::efi::TPL_CALLBACK,
+            Some(nop_notify4),
+            None,
+        )?;
+        let completion_token =
+            tcp4::CompletionToken { event: accept_event.as_raw_event(), status: Status::ABORTED };
+
+        let mut listen_token =
+            tcp4::ListenToken { completion_token, new_child_handle: crate::ptr::null_mut() };
+
+        unsafe { Self::accept_raw(self.protocol.as_ptr(), &mut listen_token) }?;
+
+        accept_event.wait()?;
+
+        let r = listen_token.completion_token.status;
+        if r.is_error() {
+            Err(status_to_io_error(r))
+        } else {
+            let child_handle = NonNull::new(listen_token.new_child_handle)
+                .ok_or(io::const_io_error!(io::ErrorKind::Other, "null child handle"))?;
+            Self::with_child_handle(self.service_binding, child_handle)
+        }
+    }
+
+    pub(crate) fn connect(&self) -> io::Result<()> {
+        let connect_event = common::Event::create(
+            r_efi::efi::EVT_NOTIFY_WAIT,
+            r_efi::efi::TPL_CALLBACK,
+            Some(nop_notify4),
+            None,
+        )?;
+        let completion_token =
+            tcp4::CompletionToken { event: connect_event.as_raw_event(), status: Status::ABORTED };
+        let mut connection_token = tcp4::ConnectionToken { completion_token };
+        unsafe { Self::connect_raw(self.protocol.as_ptr(), &mut connection_token) }?;
+        connect_event.wait()?;
+        let r = connection_token.completion_token.status;
+        if r.is_error() { Err(status_to_io_error(r)) } else { Ok(()) }
+    }
+
+    pub(crate) fn transmit(&self, buf: &[u8]) -> io::Result<usize> {
+        let buf_size = buf.len() as u32;
+        let transmit_event = common::Event::create(
+            r_efi::efi::EVT_NOTIFY_WAIT,
+            r_efi::efi::TPL_CALLBACK,
+            Some(nop_notify4),
+            None,
+        )?;
+        let completion_token =
+            tcp4::CompletionToken { event: transmit_event.as_raw_event(), status: Status::ABORTED };
+        let fragment_table = tcp4::FragmentData {
+            fragment_length: buf_size,
+            // FIXME: Probably dangerous
+            fragment_buffer: buf.as_ptr() as *mut crate::ffi::c_void,
+        };
+
+        let layout = unsafe {
+            crate::alloc::Layout::from_size_align_unchecked(
+                crate::mem::size_of::<tcp4::TransmitData>()
+                    + crate::mem::size_of::<tcp4::FragmentData>(),
+                POOL_ALIGNMENT,
+            )
+        };
+        let mut transmit_data = VariableBox::<tcp4::TransmitData>::try_new_uninit(layout)
+            .map_err(|_| common::VARIABLE_BOX_ERROR)?;
```

Comment:
Since the structure of this is statically known, I don't think this allocation is necessary.

Reply:
Well, the last member of `TransmitData` is a DST (but represented as a ZST in spec) so not sure how that can be allocated on stack.

Reply:
I have tried using const generics in `TransmitData` and `ReceiveData` to see how well that works. If everyone likes this approach here, maybe we can think about adding it to `r-efi` definition of all DSTs.